### PR TITLE
perf: use shorter reason strings

### LINKED ISSUE #3081: reduce costs of deployments to mainnet
Ideas: 

- Only deploy `NitroAdjudicator`
- reduce gasPrice
- choose deployment time wisely
- reduce size of contracts --.e.g. shorter reason strings

### PR DESCRIPTION
**Pull request by geoknee**
The test will fail for now but easily fixed.

Reduces estimated gas costs of deploying NitroAdjudicator: `2566237` becomes `2420522` => `145715` drop.

Closes #3081

### CODE REVIEW

**Comment by andrewgordstewart** [comment]
I think we should have a maintained list of error codes. Developer experience can be maintained by storing meaningful error reasons in typescript.

**Comment by geoknee** [comment]
If it is only the developer experience that benefits from the reason strings, we could run a [preprocessor](https://hardhat.org/plugins/hardhat-preprocessor.html) to strip them out of production deploys altogether?

**Comment by geoknee** [comment]
From sync with @andrewgordstewart today:

- reason strings may (or may not) be important for wallets to make decisions, too. example: I try to conclude not knowing someone else already did. If I get a revert reason 'channel already concluded' this could help me out. (There are other ways of finding this out, but they may be more complex)
- we can use the code -> verbose reason mapping suggested above to inject verbose information into our jest tests.

**Comment by geoknee** [comment]
Also we can adopt a phased approach to this

0. fixup and merge this PR for immediate gas savings with little to no loss in information
1. change to error codes because it enables more verbosity and more robust programmatic interpretation of reverts
2. strip reason strings out of production deploys if we deem that a good idea (save even more gas but lose chance to programmatically interpret reverts in client code). And hope this doesn't make bytecode verification more difficult.

File: packages/nitro-protocol/contracts/AssetHolder.sol [hunk 1]
```diff
@@ -308,7 +308,7 @@ contract AssetHolder is IAssetHolder {
             }
         }
 
-        require(affordsForDestination > 0, '_claim | guarantor affords 0 for destination');
+        require(affordsForDestination > 0, 'guarantor affords 0 for dest');
```
Comment (andrewgordstewart) [inline packages/nitro-protocol/contracts/AssetHolder.sol:311]: This is now the same string as above, worsening the debug experience.

File: packages/nitro-protocol/contracts/ForceMove.sol [hunk 2]
```diff
@@ -569,24 +563,12 @@ contract ForceMove is IForceMove {
...
             if (turnNumB < 2 * nParticipants) {
-                require(
-                    _bytesEqual(ab[1].outcome, ab[0].outcome),
-                    'InvalidTransitionError: Cannot change the default outcome during setup phase'
-                );
-                require(
-                    _bytesEqual(ab[1].appData, ab[0].appData),
-                    'InvalidTransitionError: Cannot change the appData during setup phase'
-                );
+                require(_bytesEqual(ab[1].outcome, ab[0].outcome), 'Outcome change forbidden');
```
Comment (andrewgordstewart) [inline packages/nitro-protocol/contracts/ForceMove.sol:570]: Duplicate reason string

File: packages/nitro-protocol/contracts/examples/SingleAssetPayments.sol [hunk 3]
```diff
@@ -42,11 +42,11 @@ contract SingleAssetPayments is IForceMoveApp {
         );
         require(
             assetOutcomeA.assetOutcomeType == uint8(Outcome.AssetOutcomeType.Allocation),
-            'SingleAssetPayments: outcomeA: AssetOutcomeType must be Allocation'
+            'AssetOutcome must be Allocation'
         );
         require(
             assetOutcomeB.assetOutcomeType == uint8(Outcome.AssetOutcomeType.Allocation),
-            'SingleAssetPayments: outcomeB: AssetOutcomeType must be Allocation'
+            'AssetOutcome must be Allocation'
```
Comment (andrewgordstewart) [inline packages/nitro-protocol/contracts/examples/SingleAssetPayments.sol:49]: Duplicate reason string
```suggestion
            'A.OutcomeB must be Allocation'
```

File: packages/nitro-protocol/contracts/examples/SingleAssetPayments.sol [hunk 4]
```diff
@@ -58,14 +58,8 @@ contract SingleAssetPayments is IForceMoveApp {
...
-        require(
-            allocationA.length == nParticipants,
-            'SingleAssetPayments: outcomeA: Allocation length must equal number of participants'
-        );
-        require(
-            allocationB.length == nParticipants,
-            'SingleAssetPayments: outcomeB: Allocation length must equal number of participants'
-        );
+        require(allocationA.length == nParticipants, '|Allocation|!=|participants|');
+        require(allocationB.length == nParticipants, '|Allocation|!=|participants|');
```
Comment (andrewgordstewart) [inline packages/nitro-protocol/contracts/examples/SingleAssetPayments.sol:62]:
```suggestion
        require(allocationB.length == nParticipants, '|AllocationB|!=|participants|');
```

File: packages/nitro-protocol/src/valid-transition.ts [hunk 5]
```diff
@@ -20,24 +20,12 @@ export function requireValidProtocolTransition(fromState: State, toState: State)
 
   // directly ported checks:
   if (toState.isFinal) {
-    _require(
-      isEqual(toState.outcome, fromState.outcome),
-      'InvalidTransitionError: Cannot move to a final state with a different default outcome'
-    );
+    _require(isEqual(toState.outcome, fromState.outcome), 'Outcome change forbidden');
```
Comment (andrewgordstewart) [inline packages/nitro-protocol/src/valid-transition.ts:23]: Duplicate reason?

File: packages/nitro-protocol/src/valid-transition.ts [hunk 6]
```diff
@@ -20,24 +20,12 @@ export function requireValidProtocolTransition(fromState: State, toState: State)
...
     if (toState.turnNum < 2 * toState.channel.participants.length) {
-      _require(
-        isEqual(toState.outcome, fromState.outcome),
-        'InvalidTransitionError: Cannot change the default outcome during setup phase'
-      );
-      _require(
-        toState.appData == fromState.appData,
-        'InvalidTransitionError: Cannot change the appData during setup phase'
-      );
+      _require(isEqual(toState.outcome, fromState.outcome), 'Outcome change forbidden');
```
Comment (andrewgordstewart) [inline packages/nitro-protocol/src/valid-transition.ts:27]: Duplicate reason

**Review by andrewgordstewart — APPROVED** [review]
There appear to be some duplicate reason strings, which worsens the debug experience.

I don't think this is a reason to block this PR, but I'd slightly prefer that they be unique.

**Pull request merged**